CONFIGURATION.md
## Configuration ##

Configuration of `appoptics-apm` is possible through configuration files, environment variables, and run-time
settings using the API. This covers configuration files and environment variables; the API is covered in the
api guide.

## Required Configuration ##

There is only one configuration parameter that is required, the service key. The service key is in the form
``<api token>:<service name>``.

It must be supplied using either the environment variable APPOPTICS_SERVICE_KEY or via the
`appoptics-apm-config` configuration file. If both are supplied the environment variable will be used.
If not supplied the `appoptics-apm` agent will disable itself.

## The Configuration File ##

The `appoptics-apm` default configuration file is either `appoptics-apm-config.json` or `appoptics-apm-config.js`.
It should be placed in the root directory of the project being instrumented. The file/location may be changed via the
environment variable `APPOPTICS_APM_CONFIG_NODE`. When using `APPOPTICS_APM_CONFIG_NODE` the path it specifies must
include the filename. If the file is a node module it must export a single object containing the same information that
`appoptics-apm-config.json` would; the advantage of using a node module is that the values for configuration properties
can be set programmatically.

The configuration file can supply the following properties, showing their defaults:

```
{
  serviceKey: undefined,
  enabled: true,
  hostnameAlias: undefined,
  domainPrefix: false,
  traceMode: 'enabled',
  runtimeMetrics: true,
  transactionSettings: undefined,
  insertTraceIdsIntoLogs: false,
  proxy: undefined,
  probes: {
    // probe-specific defaults. see: /docs/probe-defaults.md for details
  }
}
```

### Top Level Configuration File Properties ###

| Property Name        | Default  | Description |
| -------------------- | -------- | ----------- |
|serviceKey||As described above.|
|enabled|`true`|If set to false the `appoptics-apm` agent will disable itself.|
|proxy||Proxy that does not require authentication: `http://proxy-server:3306`. Proxy that does require Basic Authentication: `http://user:password@proxy-server:3306`.|
|hostnameAlias||A logical hostname that can be used to easily identify the host.|
|domainPrefix|`false`|Prefix transaction names with the domain name.|
|traceMode|`'enabled'`|Mode `'enabled'` will cause Appoptics to sample as many requests as possible. Mode 'disabled' will disable sampling and metrics.|
|runtimeMetrics|`true`|Collect runtime metrics characterizing the performance of node and v8|
|transactionSettings|`undefined`|An array of transactions to exclude. Each array element is an object of the form `{type: 'url', string: 'pattern', tracing: trace-setting}` or `{type: 'url', regex: /regex/, tracing: trace-setting}`. When the specified type (currently only `'url'` is implemented) matches the string or regex then tracing for that url is set to trace-setting, overriding the global traceMode. N.B. if inserting a regex into a JSON configuration file you must enter the string that is expected by the `RegExp` constructor because JSON has no representation of a `RegExp` object. `trace-setting` is either `'enabled'` or `'disabled'`.|
|ec2MetadataTimeout|`1000`|Milliseconds to wait for the ec2/openstack metadata service to respond|
|insertTraceIdsIntoLogs|`false`|Insert trace IDs into supported logging packages' output. Options are `true`, `'traced'`, `'sampledOnly'`, and `'always'`. The default, `false`, does not insert trace ids. `true` and `'traced'` insert the ID when AppOptics is tracing. `'sampledOnly'` inserts the ID when the trace is sampled. `'always'` inserts an empty trace ID value (all-zeroes) even when not tracing.|
|triggerTraceEnabled|`true`|Enable or disable the trigger-trace feature. Option values are `true` and `false`|


### Configuration File Probe Settings ###

Probes are the packages that `appoptics-apm` auto-instruments. Different types of probes have different
configuration options. See [probe defaults](./docs/probe-defaults.md) for details.

There is one particular setting for the `fs` probe that you might want to be aware of: `ignoreErrors`.
The only errors that can be ignored are node System errors (see the node [docs](https://nodejs.org/api/errors.html#errors_common_system_errors))
provided by the asynchronous functions or thrown by the synchronous functions. Note that there is only one setting for
each pair of asynchronous and synchronous functions, e.g. the `open` setting applies to both `fs.open()` and
`fs.openSync()`.

Ignoring the `ENOENT` error for the `fs.open()` and `fs.openSync()` functions is shown in the examples below. This setting
won't take effect because of the `fs` probe setting `enabled: false` but is shown for the syntax.

Probe settings in the `appoptics-apm-config` configuration file will override [probe defaults](./docs/probe-defaults.md). For example, here is how to turn off sampling for `fs`:

```
{
  "enabled": true,
  "serviceKey": "...",
  "probes": {
    "fs": {
      "enabled": false,
      "ignoreErrors": {
        "open": {
          "ENOENT": true
        }
      }
    }
  }
}
```

Bonus - the configuration file is `require`d so it can be a module (instead of JSON). A module allows
using native JavaScript with comments and avoids the need to double quote all the property names.

```
module.exports = {
  enabled: true,
  serviceKey: "...",
  // isn't this much better?
  probes: {
    fs: {
      enabled: false,
      ignoreErrors: {
        open: {
          ENOENT: true,
        }
      }
    }
  }
}
```

#### Environment Variables ####

Environment variables with illegal values will generate a warning message and revert to their default values.
Variables accepting `'true'` and `'false'` as options care case-insensitive.

These environment variables may be set:

| Variable Name        | Default  | Description |
| -------------------- | -------- | ----------- |
|APPOPTICS_ENABLED|`true`|The agent is enabled. Set to `false` to disable.|
|APPOPTICS_LOG_SETTINGS|`'error,warn'`|Categories to log. If set this takes precedence over the deprecated DEBUG environment variable. If set to empty it suppresses all agent log messages.|
|APPOPTICS_PROXY||Proxy that does not require authentication: `http://proxy-server:3306`. Proxy that does require Basic Authentication: `http://user:password@proxy-server:3306`.|
|APPOPTICS_APM_CONFIG_NODE|`'$PWD/appoptics-apm'`|The location of the configuration file.|
|APPOPTICS_DEBUG_LEVEL|`'2'`|Logging level for low-level library. Higher numbers get more logging. Possible values: 1 to 6|
|APPOPTICS_TRIGGER_TRACE_ENABLED|`'true'`|Enable the trigger-trace feature. Options are `'true'`, `'false'`|
|APPOPTICS_RUNTIME_METRICS|`'true'`|Enable runtime metrics. Options are `'true'`, `'false'`|
|APPOPTICS_REPORTER|`'ssl'`|The reporter that will be used throughout the runtime of the app. Possible values: `'ssl'`, `'udp'`. This is used for testing.|
|APPOPTICS_COLLECTOR|`'collector.appoptics.com:443'`|SSL collector endpoint address and port. This is typically changed only for testing.|
|APPOPTICS_TRUSTEDPATH|built-in|Path to the certificate used to verify the collector endpoint. Used only for testing.|


guides/configuration.md
## Configuration ##

Configuration of `appoptics-apm` is possible through configuration files, environment variables, and run-time
settings using the API. This covers configuration files and environment variables; the API is covered in the
api guide.

## Required Configuration ##

There is only one configuration parameter that is required, the service key. The service key is in the form
``<api token>:<service name>``.

It must be supplied using either the environment variable APPOPTICS_SERVICE_KEY or via the
`appoptics-apm-config` configuration file. If both are supplied the environment variable will be used.
If not supplied the `appoptics-apm` agent will disable itself.

## The Configuration File ##

The `appoptics-apm` default configuration file is either `appoptics-apm-config.json` or `appoptics-apm-config.js`.
It should be placed in the root directory of the project being instrumented. The file/location may be changed via the
environment variable `APPOPTICS_APM_CONFIG_NODE`. When using `APPOPTICS_APM_CONFIG_NODE` the path it specifies must
include the filename. If the file is a node module it must export a single object containing the same information that
`appoptics-apm-config.json` would; the advantage of using a node module is that the values for configuration properties
can be set programmatically.

The configuration file can supply the following properties, showing their defaults:

```
{
  serviceKey: undefined,
  enabled: true,
  hostnameAlias: undefined,
  domainPrefix: false,
  ignoreConflicts: false,
  traceMode: 'enabled',
  runtimeMetrics: true,
  transactionSettings: undefined,
  insertTraceIdsIntoLogs: false,
  insertTraceIdsIntoMorgan: false,
  createTraceIdsToken: undefined,
  proxy: undefined,
  probes: {
    // probe-specific defaults. see lib/probe-defaults.js for details
  }
}
```

### Top Level Configuration File Properties ###

| Property Name        | Default  | Description |
| -------------------- | -------- | ----------- |
|serviceKey||As described above.|
|enabled|`true`|If set to false the `appoptics-apm` agent will disable itself.|
|proxy||Proxy that does not require authentication: `http://proxy-server:3306`. Proxy that does require Basic Authentication: `http://user:password@proxy-server:3306`.|
|hostnameAlias||A logical hostname that can be used to easily identify the host.|
|domainPrefix|`false`|Prefix transaction names with the domain name.|
|ignoreConflicts|`false`|Appoptics will disable itself when conflicting APM products are loaded unless this is set to `true`.|
|traceMode|`'enabled'`|Mode `'enabled'` will cause Appoptics to sample as many requests as possible. Mode 'disabled' will disable sampling and metrics.|
|runtimeMetrics|`true`|Collect runtime metrics characterizing the performance of node and v8|
|transactionSettings|`undefined`|An array of transactions to exclude. Each array element is an object of the form `{type: 'url', string: 'pattern', tracing: trace-setting}` or `{type: 'url', regex: /regex/, tracing: trace-setting}`. When the specified type (currently only `'url'` is implemented) matches the string or regex then tracing for that url is set to trace-setting, overriding the global traceMode. N.B. if inserting a regex into a JSON configuration file you must enter the string that is expected by the `RegExp` constructor because JSON has no representation of a `RegExp` object. `trace-setting` is either `'enabled'` or `'disabled'`.|
|ec2MetadataTimeout|`1000`|Milliseconds to wait for the ec2/openstack metadata service to respond|
|insertTraceIdsIntoLogs|`false`|Insert trace IDs into supported logging packages' output. Options are `true`, `'traced'`, `'sampledOnly'`, and `'always'`. The default, `false`, does not insert trace ids. `true` and `'traced'` insert the ID when AppOptics is tracing. `'sampledOnly'` inserts the ID when the trace is sampled. `'always'` inserts an empty trace ID value (all-zeroes) even when not tracing.|
|insertTraceIdsIntoMorgan|`false`|Append trace IDs to morgan log lines. Because morgan does not output JSON the morgan formats must be modified to enable the trace IDs to be appended. This is a more invasive approach than inserting a property in a JSON object so it requires this explicit setting. The options are the same as for `insertTraceIdsIntoLogs`.|
|createTraceIdsToken|`undefined`|Create a token that can be used in a logging package's format string. If set to `'morgan'` the token `ao-auto-trace-id` will be created and `:ao-auto-trace-id` can be used in morgan format strings.|
|triggerTraceEnabled|`true`|Enable or disable the trigger-trace feature. Option values are `true` and `false`|


### Configuration File Probe Settings ###

Probes are the packages that `appoptics-apm` auto-instruments. Different types of probes have different
configuration options. See `lib/probe-defaults.js` for details.

There is one particular setting for the `fs` probe that you might want to be aware of: `ignoreErrors`.
The only errors that can be ignored are node System errors (see the node [docs](https://nodejs.org/api/errors.html#errors_common_system_errors))
provided by the asynchronous functions or thrown by the synchronous functions. Note that there is only one setting for
each pair of asynchronous and synchronous functions, e.g. the `open` setting applies to both `fs.open()` and
`fs.openSync()`.

Ignoring the `ENOENT` error for the `fs.open()` and `fs.openSync()` functions is shown in the examples below. This setting
won't take effect because of the `fs` probe setting `enabled: false` but is shown for the syntax.

Probe settings in the `appoptics-apm-config` configuration file will override those in `probe-defaults.js`, so the safest
approach to changing an option is to add it to `appoptics-apm-config.json`. For example, here is how to turn off sampling
for `fs`:

```
{
  "enabled": true,
  "serviceKey": "...",
  "probes": {
    "fs": {
      "enabled": false,
      "ignoreErrors": {
        "open": {
          "ENOENT": true
        }
      }
    }
  }
}
```

Bonus - the configuration file is `require`d so it can be a module (instead of JSON). A module allows
using native JavaScript with comments and avoids the need to double quote all the property names.

```
module.exports = {
  enabled: true,
  serviceKey: "...",
  // isn't this much better?
  probes: {
    fs: {
      enabled: false,
      ignoreErrors: {
        open: {
          ENOENT: true,
        }
      }
    }
  }
}
```

#### Environment Variables ####

Environment variables with illegal values will generate a warning message and revert to their default values.
Variables accepting `'true'` and `'false'` as options care case-insensitive.

These environment variables may be set:

| Variable Name        | Default  | Description |
| -------------------- | -------- | ----------- |
|APPOPTICS_ENABLED|`true`|The agent is enabled. Set to `false` to disable.|
|APPOPTICS_LOG_SETTINGS|`'error,warn'`|Categories to log. If set this takes precedence over the deprecated DEBUG environment variable. If set to empty it suppresses all agent log messages.|
|APPOPTICS_PROXY||Proxy that does not require authentication: `http://proxy-server:3306`. Proxy that does require Basic Authentication: `http://user:password@proxy-server:3306`.|
|APPOPTICS_APM_CONFIG_NODE|`'$PWD/appoptics-apm'`|The location of the configuration file.|
|APPOPTICS_DEBUG_LEVEL|`'2'`|Logging level for low-level library. Higher numbers get more logging. Possible values: 1 to 6|
|APPOPTICS_TRIGGER_TRACE_ENABLED|`'true'`|Enable the trigger-trace feature. Options are `'true'`, `'false'`|
|APPOPTICS_RUNTIME_METRICS|`'true'`|Enable runtime metrics. Options are `'true'`, `'false'`|
|APPOPTICS_REPORTER|`'ssl'`|The reporter that will be used throughout the runtime of the app. Possible values: `'ssl'`, `'udp'`. This is used for testing.|
|APPOPTICS_COLLECTOR|`'collector.appoptics.com:443'`|SSL collector endpoint address and port. This is typically changed only for testing.|
|APPOPTICS_TRUSTEDPATH|built-in|Path to the certificate used to verify the collector endpoint. Used only for testing.|

Deprecated environment variables. They will be removed in the future:

| Deprecated Name      | Default  | Description |
| -------------------- | -------- | ----------- |
|DEBUG|`'appoptics:error,appoptics:warn'`|This has been replaced with APPOPTICS_LOG_SETTINGS so the 'appoptics:' prefix does not need to be entered for each setting.|
|APPOPTICS_TRIGGER_TRACE|`'enable'`|Enable or disable the trigger-trace feature. Options are `'enable'`, `'disable'`|




CHANGELOG.md
## appoptics-apm changelog

### v6.0.0

Features and bug fixes
- add per-url transaction filtering
- route handling code and middleware for express, koa, and restify are now traced.
- use APPOPTICS_LOG_SETTINGS to set log levels; using the DEBUG environment variable is deprecated.
    - `export APPOPTICS_LOG_SETTINGS=error,warn` as opposed to `export DEBUG=appoptics:error,appoptics:warn`
- issue explicit log warning if disabled by config.
- fix koa-router probe to work with multiple middleware arguments.
- fix incorrect oboe library version reporting in init message.

Breaking changes
- all breaking changes are in the API. See `guides/migration-5to6.md` for details.

#### v6.1.0

Features and bug fixes
- enable inserting trace IDs into logs automatically (pino, winston)
- new API function, `ao.getFormattedTraceId()` to get trace IDs using code.
- new API function, `ao.sendMetric()` for sending custom metrics.
- fix config file not found bug.

#### v6.2.0

Features
- enable inserting trace IDs into logs automatically (bunyan)
- `insertTraceIdsIntoLogs` options expanded to `false`, `true`, `'traced'`, `'sampledOnly'`, `'always'`

#### v6.3.0

Features and bug fixes
- add `insertTraceIdsIntoMorgan` to enable appending `ao.traceId=...` format morgan's text output.
- add `createTraceIdsToken`. Set to `'morgan'` to have the token `ao-auto-trace-id` token created. Use as `:ao-auto-trace-id` in morgan formats.
- set config to service key that was used.
- fix aws-sdk bad signature error on transaction retry.

#### v6.4.0

Features and bug fixes
- add support for SolarWinds Tokens (swoken)
- incorporate appoptics-bindings v6.2.1 with custom metrics fix

### v6.5.0

Features and bug fixes
- internal environment variable reorganization
- fix corrupted histogram in node 11.10.0+.
- support @hapi scoped packages hapi & vision.

### v6.5.1

Features and bug fixes
- use oboe v5.1.1, correcting histogram memory leak

### v6.6.0

Features and bug fixes
- context management simplification + bindings 6.4.0
- handle more boolean environment variable variations.
- propagate FilePath KV value correctly in edge case
- add log setting 'event:create', cassandra-driver now always uses 'patching'

### v6.7.0

Features and bug fixes
- trigger-trace
- appoptics-bindings 7.0.0
- loosen bind-emitter checks
- warn only when unknown environment variables

### v6.7.1

Bug fixes
- make TemplateLanguage KV consistent - never include leading dot.
- bind finalizer in instrumentHttp.
- fix zlib bind emitter message.

### v6.8.0

Features
- fetch container information in Azure App Service environment.

### v6.9.0

Features
- new `mongodb` probe for versions >= 3.3.0 as it now longer uses `mongodb-core`.
- allow the `fs` probe to ignore specific errors.
- `patching` log setting logs the version of patched modules.
- disable `mongodb-core` versions < 3 for node versions > 11.15.0 due to v8 memory leak.

Bug fixes
- supply a default `TransactionName` when none is present.
- log only one message for `getTraceSettings()` errors.
- log only one message for each incorrect KV pair.
- reset both count and time windows when a debounced log message is issued.
- add missing `FileDescriptor` to exit events.

### v7.0.0

Features
- add forceNewTrace option to `startOrContinueTrace()`
- support brotli compressin in `zlib`
- support restify v7+
- use `ace-context`; remove `continuation-local-storage`

Bug fixes
- force new context on inbound HTTP requests.
- fix http `RemoteURL` KV when search/query is present
- exit http/s client spans correctly on socket errors
- exit http/s client spans on upgrade events
- don't add undefined `Database` KV
- use new http_parser values if present

### v7.1.0

Features
- runtime metrics
- new function `sendMetrics()`; deprecate `sendMetric()`
- requires `appoptics-bindings@9`

### v7.1.1

Bug fix
- fix undefined `reporter` in `sendMetrics()`

### v8.0.0

Features
- minimize span/event creation for unsampled traces
- support HTTP proxy configuration
- improve run span logic.
- debounce metrics send error logging
- improve logging consistency
- implement messaging for host-requested soft disable

Breaking changes
- removed `Span.last` and `Event.last` - use `ao.lastSpan` and `ao.lastEvent`
- removed `span.exitWithError()` - use `span.exitCheckingError()`
- `aob.Reporter.isReadyToSample()` is now `aob.isReadyToSample()`
- `aob.Metadata` no longer exists. see `guides/migration.md` if you're using this low level class.
- removed `%m` custom log format for `Metadata` objects.

### v8.0.1

Bug fixes
- support cassandra-driver > v4.4.0
- document proxy in configuration guide.

### v8.1.0

Features
- http/https probe configuration property to specify header to be used for ClientIP KV
- don't read instrumented package versions more than once.

Bug fix
- don't log an error when `req.socket.remoteAddress` is undefined.

### v9.0.0

Features
- lambda support
- capture agent version in init message

Breaking change
- removed `ao.serviceKey` property.

Bug fix
- change config file name `appoptics-apm.{js|json}` => `appoptics-apm-config.{js|json}`

### v10.0.0

Features
- `@appoptics/apm-bindings` replaces `appoptics-bindings`. `@appoptics/apm-bindings` uses `node-pre-gyp` so that users of LTS versions of node do not need to have the c/c++ build chain installed.
- if `APPOPTICS_LOG_SETTINGS` is empty suppress all logging
- use shimmer instead of ximmer
- add `span.runPromise()` - native async span runner
- updated to Apache 2.0 license

Breaking change
- Removed `ao.serviceKey`
- Configuration file name-change

Bug fix
- pass options to koa constructor
- check if module needs patching before doing builtin lookup
- set txname when not set by frameworks

### v10.0.1

Bug fix
- avoid triggering non-existent property warning during module.exports on Node 14

### v10.1.0

Maintenance
- Published using GitHub Actions
- Supports Node 16
- Added node-pre-gyp distributable binaries for Node 16
- Using oboe c-lib v10.1.1

env.sh
#!/bin/bash
ARG=$1
PARAM=$2

AO_TOKEN=${AO_TOKEN_STG:-$AO_TOKEN_PROD}

if [ -z "$AO_TOKEN" ]; then
    echo "AO_TOKEN_PROD or AO_TOKEN_STG must be defined and contain a valid SolarWinds token"
    echo "for accessing a collector."
    return
fi

if [ "$AO_TOKEN" = "$AO_TOKEN_PROD" ]; then
    APPOPTICS_COLLECTOR=collector.appoptics.com
else
    APPOPTICS_COLLECTOR=collector-stg.appoptics.com
fi

global_path="$NVM_DIR/versions/node/$(node -v)/lib/node_modules"
if [ -d "$global_path" ] && [[ ":$PATH:" != *":$1:"* ]]; then
    PATH="${PATH:+"$PATH:"}$global_path"
elif [ ! -d "$global_path" ]; then
    echo "$global_path is not a directory, skipping"
fi

if [[ -z "$ARG" ]]; then
    echo "source this script with an argument of docker, docker-scribe, bash,"
    echo "bash-testing"
    echo ""
    echo "docker defines variables for running tests in the docker environment."
    echo "docker-scribe does the same but with the scribe collector instead of java"
    echo "  collector."
    echo "bash defines variables to run at a native command prompt. N.B. databases"
    echo "  will not necessarily be defined."
    echo "bash-testing defines the variables to test from the bash prompt but with docker"
    echo "  containers present."
    echo
    echo "you may also use the argument debug to define additional"
    echo "authentication and package, or tcpdump to get help on tcpdump"
elif [[ "$ARG" = "key" ]]; then
    if [[ -z "$PARAM" ]]; then
        echo "defining the service key requires a service name argument"
        return
    fi
    export APPOPTICS_SERVICE_KEY=${AO_TOKEN}:$PARAM
    echo "set APPOPTICS_SERVICE_KEY=$APPOPTICS_SERVICE_KEY"
elif [[ "$ARG" = "add-bin" ]]; then
    # add ./node_modules/.bin to PATH
    [[ ":$PATH" != *":$PWD/node_modules/.bin"* ]] && PATH="${PATH}:$PWD/node_modules/.bin"
elif [[ "$ARG" = "docker-java" ]]; then
    export APPOPTICS_TRUSTEDPATH=/appoptics/test/certs/java-collector.crt
    export APPOPTICS_COLLECTOR=java-collector:12222
    export APPOPTICS_SERVICE_KEY=${AO_TOKEN}:ao-node-test-docker
    export APPOPTICS_REPORTER=ssl
elif [[ "$ARG" = "docker-scribe" ]]; then
    export APPOPTICS_TRUSTEDPATH=/appoptics/test/certs/scribe-collector.crt
    export APPOPTICS_COLLECTOR=scribe-collector:4444
    export APPOPTICS_SERVICE_KEY=${AO_TOKEN}:ao-node-test-docker
    export APPOPTICS_REPORTER=ssl
elif [[ "$ARG" = "bash" ]]; then
    # this is used primarily for manual interactive testing so doesn't connect to a
    # real collector as most tests will fail if it does.
    [[ ":$PATH" != *":$PWD/node_modules/.bin"* ]] && PATH="${PATH}:$PWD/node_modules/.bin"

    # set logging that should be seen during testing.
    export APPOPTICS_LOG_SETTINGS=error,warn,patching,bind,debug

    export APPOPTICS_COLLECTOR=localhost:7832
    export APPOPTICS_SERVICE_KEY=${AO_TOKEN}:ao-node-test
    export APPOPTICS_REPORTER=udp

    # and the buckets need to be adjusted if using UDP because the defaults
    # result in lost messages without notification.
    # NOTE - no longer required after oboe changes.
    #export APPOPTICS_TOKEN_BUCKET_CAPACITY=1000
    #export APPOPTICS_TOKEN_BUCKET_RATE=1000

    # the following are used primarily to run the full appoptics-apm test suite.
    # presumes docker containers are running and their ports are addressable
    # as localhost. the port overrides (e.g., AO_TEST_MYSQL_HOST_PORT) allow
    # existing local copies of the database that run on the standard port
    # numbers to be used for testing (and not conflict with the port docker
    # would normally expose).
    export AO_TEST_CASSANDRA_2_2=localhost:9042
    export AO_TEST_MEMCACHED_1_4=localhost:11211
    export AO_TEST_MONGODB_2_4=localhost:27016
    export AO_TEST_MONGODB_2_6=localhost:${AO_TEST_MONGO_2_6_HOST_PORT:-27017}
    export AO_TEST_MONGODB_3_0=localhost:27018
    #export AO_TEST_MONGODB_SET=localhost:30001,localhost:30002,localhost:30003
    DTS=ec2-52-7-124-5.compute-1.amazonaws.com
    export AO_TEST_MONGODB_SET=$DTS:10301,$DTS:10302,$DTS:10303

    # enable mysql to run with different port
    export AO_TEST_MYSQL=localhost:${AO_TEST_MYSQL_HOST_PORT:-3306}
    # if different port then use default user/password settings
    export AO_TEST_MYSQL_USERNAME=${AO_TEST_MYSQL_HOST_USERNAME:root}
    export AO_TEST_MYSQL_PASSWORD=${AO_TEST_MYSQL_HOST_PASSWORD+}
    # defaults should be fine.
    #export AO_TEST_POSTGRES_USER=postgres
    #export AO_TEST_POSTGRES_PASSWORD=
    export AO_TEST_POSTGRES=localhost:5432
    export AO_TEST_RABBITMQ_3_5=localhost:5672
    export AO_TEST_REDIS_3_0=localhost:6379
    # the tedious probe tests SQL Server.
    export AO_TEST_SQLSERVER_EX=localhost:1433
    export AO_TEST_SQLSERVER_USERNAME=sa
    export AO_TEST_SQLSERVER_PASSWORD=whatyouknowaboutthat
elif [[ "$ARG" = "stg" ]]; then
    export APPOPTICS_COLLECTOR=collector-stg.appoptics.com
    export APPOPTICS_SERVICE_KEY=${AO_TOKEN_STG}:ao-node-test
    export APPOPTICS_REPORTER=ssl
elif [[ "$ARG" = "prod" ]]; then
    export APPOPTICS_COLLECTOR=collector.appoptics.com
    export APPOPTICS_SERVICE_KEY=${AO_TOKEN_PROD}:ao-node-test
    export APPOPTICS_REPORTER=ssl
elif [[ "$ARG" = "debug" ]]; then
    # this section is more for reference than anything else.
    # LEVEL 2 is most of what you want to see. 6 (highest) is too much.
    export APPOPTICS_DEBUG_LEVEL=2
    # see lib/loggers.js for all the options
    export APPOPTICS_LOG_SETTINGS=error,warn,debug,patching,bind
    #export APPOPTICS_TOKEN_BUCKET_CAPACITY=1000
    #export APPOPTICS_TOKEN_BUCKET_RATE=1000
elif [[ "$ARG" = "help" ]]; then
    echo "try"
    echo "    $ sudo tcpdump -i lo -n udp port 7832"
    echo "to watch the UDP traffic"
    #sudo tcpdump -i lo -n udp port 7832
else
    echo "ERROR $ARG invalid"
fi

return




